Cabal/src/Distribution/Simple/Build.hs

```diff
@@ -434,52 +433,36 @@ generateCode codeGens nm pdesc bi lbi clbi verbosity = do
 -- information.
 addExtraCSources :: BuildInfo -> [FilePath] -> BuildInfo
 addExtraCSources bi extras = bi { cSources = new }
-  where new = Set.toList $ old `Set.union` exs
-        old = Set.fromList $ cSources bi
-        exs = Set.fromList extras
-
+  where new = ordNub (extras ++ cSources bi)
 
 -- | Add extra C++ sources generated by preprocessing to build
 -- information.
 addExtraCxxSources :: BuildInfo -> [FilePath] -> BuildInfo
 addExtraCxxSources bi extras = bi { cxxSources = new }
-  where new = Set.toList $ old `Set.union` exs
-        old = Set.fromList $ cxxSources bi
-        exs = Set.fromList extras
-
+  where new = ordNub (extras ++ cxxSources bi)
 
 -- | Add extra C-- sources generated by preprocessing to build
 -- information.
 addExtraCmmSources :: BuildInfo -> [FilePath] -> BuildInfo
 addExtraCmmSources bi extras = bi { cmmSources = new }
-  where new = Set.toList $ old `Set.union` exs
-        old = Set.fromList $ cmmSources bi
-        exs = Set.fromList extras
-
+  where new = ordNub (extras ++ cmmSources bi)
 
 -- | Add extra ASM sources generated by preprocessing to build
 -- information.
 addExtraAsmSources :: BuildInfo -> [FilePath] -> BuildInfo
 addExtraAsmSources bi extras = bi { asmSources = new }
-  where new = Set.toList $ old `Set.union` exs
-        old = Set.fromList $ asmSources bi
-        exs = Set.fromList extras
+  where new = ordNub (extras ++ asmSources bi)
 
 -- | Add extra HS modules generated by preprocessing to build
 -- information.
 addExtraOtherModules :: BuildInfo -> [ModuleName.ModuleName] -> BuildInfo
 addExtraOtherModules bi extras = bi { otherModules = new }
-  where new = Set.toList $ old `Set.union` exs
-        old = Set.fromList $ otherModules bi
-        exs = Set.fromList extras
+  where new = ordNub (extras ++ otherModules bi)
 
 -- | Add extra source dir for generated modules.
 addSrcDir :: BuildInfo -> FilePath -> BuildInfo
 addSrcDir bi extra = bi { hsSourceDirs = new }
-  where new = Set.toList $ old `Set.union` ex
-        old = Set.fromList $ hsSourceDirs bi
-        ex  = Set.fromList [unsafeMakeSymbolicPath extra] -- TODO
```

Comment:
What happened to the `TODO`?

Reply:
yeah, neither do I, unfortunately. I hoped you fixed it somehow :D

Reply:
What it meant I have a guess: maybe the author of TODO wished there was no "unsafe"?